### DISCUSSION TITLE
Defining header on the request body content?

### DISCUSSION THREAD

**Discussion by konst-sh** [fsprojects/FsHttp#120] [Q&A] [answered]
I'm trying to rewrite some code from the native HttpClient to this builder. The code using Content-Type header on the body of the GET request (it was actually supposed to be in the request header, but at some moment I found that it was impossible to set Content-Type header on the GET request itself in the .NET so I had to make en empty string content for the request and attach the header to it).
Was not able to find anything similar to setting content headers in the documentation here, so is this possible on the at all?

Here is the snippet that originally worked for me:
```
    let reqMessage = new System.Net.Http.HttpRequestMessage(System.Net.Http.HttpMethod.Get , url)
    [
        "Accept",           "application/vnd.onshape.v2+json;charset=UTF-8;qs=0.2"
        "On-Nonce",         nonce
        "Authorization",    signature
        "Content-Type",     contentType
    ]
    |>List.iter ( fun (h, value) -> reqMessage.Headers.TryAddWithoutValidation(h, value) |> ignore )

    reqMessage.Content <- new System.Net.Http.StringContent("")
    reqMessage.Content.Headers.ContentType <- System.Net.Http.Headers.MediaTypeHeaderValue(contentType)

    let resStr = 
        task {
            use client = new System.Net.Http.HttpClient()
            let! res = client.SendAsync(reqMessage)
            return! res.Content.ReadAsStringAsync()
        }
    resStr.Result
```
and this fails at compile time with the error:
`System.InvalidOperationException: Misused header name, 'Content-Type'. Make sure request headers are used with HttpRequestMessage, response headers with HttpResponseMessage, and content headers with HttpContent objects.`
```
    http {
        GET url
        header "Accept"          "application/vnd.onshape.v2+json;charset=UTF-8;qs=0.2"
        header "On-Nonce"         nonce
        header "Authorization"    signature
        header "Date"             (dateTime.ToString())
        header "Content-Type"     contentType
    }
    |> Request.send
```

**Comment by SchlenkR** [comment, marked as answer]
Hi @konst-sh,

some month ago you wrote; here's my reply :)

* there are well-known headers, which can be used in a safe way, instead of using strings as header keys.
* `ContentType` is a heady related to the body, so it has to be defined in the `body` section of the request. **Better:** you use one of the available custom operations like `json` or `jsonSerialize`, which sets the content type automatically.

```fsharp
http {
    GET "url"

    Accept "application/vnd.onshape.v2+json;charset=UTF-8;qs=0.2"
    header "OnNonce" "nonce"
    AuthorizationBearer "token"
    Date DateTime.Now

    body
    ContentType "application/json"
}
|> Request.send
```

**Reply by konst-sh** [reply]
Hi, thank you, I was able to find that myself almost at the same day)
